src/ucs/datastruct/lru.h

```diff
@@ -108,25 +102,26 @@ static UCS_F_ALWAYS_INLINE void ucs_lru_put(ucs_lru_h lru, void *key)
 
     if (ucs_likely(ret == UCS_KH_PUT_KEY_PRESENT)) {
         ucs_list_del(&elem->list);
-    } else if (kh_size(&lru->hash) == lru->capacity) {
+    } else if (kh_size(&lru->hash) > lru->capacity) {
         ucs_lru_pop(lru);
     }
 
     ucs_list_add_head(&lru->list, &elem->list);
 }
 
+
+#define ucs_lru_next_key(_elem) \
+    &ucs_container_of(_elem->list.next, ucs_lru_element_t, list)->key
+
+
 /**
- * @brief Get a list of all elements in the cache.
- *
- * @param [in]  lru     Handle to the LRU cache.
- * @param [in]  cb      Callback function to be called for each element in the
- *                      cache.
- * @param [in]  opaque  Pointer to user data object. Will be passed to the
- *                      callback function.
- *
- * @return UCS_OK if successful, or error code otherwise.
+ * Iterate over elements of the LRU.
  */
-ucs_status_t
-ucs_lru_get_all(const ucs_lru_h lru, ucs_lru_get_cb_t cb, void *opaque);
+#define ucs_lru_for_each(_elem, _lru) \
+    for (_elem = ucs_lru_next_key(_lru); \

```

Comment:
how is it possible to call ucs_lru_next_key on the lru struct?

Reply:
it's the list head (lru->list)

Reply:
so it works because the field name is the same?